### ISSUE TITLE
--gulpfile and --cwd not working as expected

### ISSUE THREAD

**Issue by trusktr** [gulpjs/gulp#2126]
Based on https://github.com/gulpjs/gulp/issues/122, I'm trying to use `--gulpfile` and `--cwd` without luck.

> Are you trying to look up folders or down folders for a gulpfile?

I'm trying to look up folders.

I have this structure:

```
project # I want to run gulp in here
  package.json # has gulp dependency
  node_modules # local gulp install here
  src/ # source files
otherFolder
  gulpfile.js 
```

where `project` is the folder I'm in while at the command line, then I'm running this:

```sh
gulp --cwd ./ --gulpfile ../otherFolder/gulpfile.js
```

but it gives me this error:

```
trusktr@DESKTOP-7H13Q4E:/path/to/project$ gulp --cwd ./ --gulpfile ../otherFolder/gulpfile.js
[14:07:29] Local gulp not found in /path/to/project
[14:07:29] Try running: npm install gulp
```

although, `gulp` **_is_** installed in the `project`:

```
trusktr@DESKTOP-7H13Q4E:/path/to/project$ ls -ld node_modules/gulp/
drwxrwxrwx 0 root root 4096 Feb 11 12:41 node_modules/gulp/
```

What I'm _really_ trying to do is use [Builder](http://formidable.com/open-source/builder) to share common configuration across projects and packages, so I'd like to ship a `gulpfile.js` inside an NPM package and share it (this is called an "archetype" in Builder).

If I link my "archetype" package into my project, so that it will be inside of `node_modules`, and therefore I am _looking down_ for the gulpfile, still no luck:

```
trusktr@DESKTOP-7H13Q4E:/path/to/project$ gulp --cwd ./ --gulpfile node_modules/builder-js-package/config/gulpfile.js
module.js:540
    throw err;
    ^

Error: Cannot find module 'gulp'
    at Function.Module._resolveFilename (module.js:538:15)
    at Function.Module._load (module.js:468:25)
    at Module.require (module.js:587:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/mnt/d/src/trusktr+builder-js-package/config/gulpfile.js:1:76)
    at Module._compile (module.js:643:30)
    at Object.Module._extensions..js (module.js:654:10)
    at Module.load (module.js:556:32)
    at tryModuleLoad (module.js:499:12)
    at Function.Module._load (module.js:491:3)
```

where `builder-js-package` is the name of my "archetype".

Basically, `--gulpfile` and `--cwd` just doesn't seem to work like expected.

**Comment by phated** [comment]
Don't believe this is possible because no one ever submitted a patch for https://github.com/js-cli/js-liftoff/issues/55.

I'm not particularly fond of creating separate gulpfiles in other packages and using them - instead, you should create a Custom Registry for gulp 4 and share that.

**Comment by phated** [comment]
Pinging @sttk to make sure I'm correct on that first part.

**Comment by trusktr** [comment]
> Don't believe this is possible

So then what _is_ possible? The [gulp-cli README](https://github.com/gulpjs/gulp-cli) doesn't really explain what's possible with `--gulpfile` and `--cwd`.

> instead, you should create a Custom Registry for gulp 4 and share that.

If I understand correctly, the custom registry would merely contain tasks bundled up in a container. If I were then to use a registry, then I'd need to ship this registry in another package (`builder-js-package` in the above example). How is this going to make it work? I don't want projects all importing and using the custom registry manually, I want my `builder-js-package` package to contain gulp and all build steps, and there (ideally) won't be any trace of `gulp` in the projects where I'm using `builder-js-package`.

**Comment by phated** [comment]
That's a bad pattern...

**Comment by trusktr** [comment]
In other words, using a custom registry, the only thing that will change is the gulpfile will be importing tasks from another file:

```js
var companyTasks = require('myCompanyTasksRegistry');

gulp.registry(companyTasks);
```

but that won't magically make it work the way I'm trying with `--gulpfile` and `--cwd`, where gulp is not run directly in the project, but ran by another package. It would be great to be able to run gulp from somewhere else, and for it to simply work.

For example, suppose `gulp` is installed _**only**_ globally, and there's a gulpfile somewhere in the system. I should be able to go into any project (assuming projects have a common file structure) and just do

```
cd project
gulp --cwd . --gulpfile /path/to/gulpfile
```

If that doesn't work, then the whole intuitive nature of what those flags represents seems to be broken. Is it planned to be fixed?

**Comment by trusktr** [comment]
> That's a bad pattern...

How is it a bad pattern?

**Comment by phated** [comment]
It'll never work to have gulp only installed globally. Node doesn't work that way

**Comment by trusktr** [comment]
You're just being opinionated. _You_ don't have to do it that way.

**Comment by phated** [comment]
Not only that, but gulp shouldn't be installed globally, only gulp-cli

**Comment by phated** [comment]
I'm not being opinionated on how node's `require` works... `require('blah')` doesn't require from global.

**Comment by trusktr** [comment]
That's why, f.e., `webpack.config.js` isn't required to `require` anything. All it needs to do is export a config object. It can be `webpack.config.json` even. Then, a global `webpack` reads it just fine.

**Comment by trusktr** [comment]
Anywho, I'm not saying I want `gulp` to be global, I want to use `gulp` from another package, to manage all my other packages.

**Comment by phated** [comment]
> That's why, f.e., webpack.config.js isn't required to require anything. All it needs to do is export a config object. It can be webpack.config.json even. Then, a global webpack reads it just fine.

I think you are in the wrong project...  When you start comparing gulp to webpack, you are wayyyyyyyy off base.

I'm just going to close this now.  Please don't keep discussing here so I don't need to lock this.  I've already pointed you to the issue related to --cwd + --gulpfile usage and everything else should be directed to StackOverflow.

**Comment by phated** [comment]
@trusktr sounds good!
